cmd/prometheus/main.go

```diff
@@ -846,6 +817,37 @@ func main() {
 	)
 
 	if !agentMode {
+		notifierManager = notifier.NewManager(&cfg.notifier, logger.With("component", "notifier"))
+
+		// Kubernetes client metrics are used by Kubernetes SD.
+		// They are registered here in the main function, because SD mechanisms
+		// can only register metrics specific to a SD instance.
+		// Kubernetes client metrics are the same for the whole process -
+		// they are not specific to an SD instance.
+		err = discovery.RegisterK8sClientMetricsWithPrometheus(prometheus.DefaultRegisterer)
+		if err != nil {
+			logger.Error("failed to register Kubernetes client metrics", "err", err)
+			os.Exit(1)
+		}
+
+		sdMetrics, err := discovery.CreateAndRegisterSDMetrics(prometheus.DefaultRegisterer)
+		if err != nil {
+			logger.Error("failed to register service discovery metrics", "err", err)
+			os.Exit(1)
+		}
+
+		discoveryManagerScrape = discovery.NewManager(ctxScrape, logger.With("component", "discovery manager scrape"), prometheus.DefaultRegisterer, sdMetrics, discovery.Name("scrape"))
```

Comment:
The ordering of start-up and shut-down between components may be important, so unless you have some argument that you know for sure it's ok to change, please leave the ordering the way it was.